pkg/chunk/series_store.go

```diff
@@ -111,12 +111,54 @@ func (c *seriesStore) Get(ctx context.Context, from, through model.Time, allMatc
 		return nil, err
 	}
 
+	chks, fetchers, err := c.GetChunkRefs(ctx, from, through, allMatchers...)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(chks) == 0 {
+		// Shortcut
+		return nil, nil
+	}
+
+	chunks := chks[0]
+	fetcher := fetchers[0]
+	// Protect ourselves against OOMing.
+	maxChunksPerQuery := c.limits.MaxChunksPerQuery(userID)
+	if maxChunksPerQuery > 0 && len(chunks) > maxChunksPerQuery {
+		err := httpgrpc.Errorf(http.StatusBadRequest, "Query %v fetched too many chunks (%d > %d)", allMatchers, len(chunks), maxChunksPerQuery)
+		level.Error(log).Log("err", err)
+		return nil, err
+	}
+
+	// Now fetch the actual chunk data from Memcache / S3
+	keys := keysFromChunks(chunks)
+	allChunks, err := fetcher.FetchChunks(ctx, chunks, keys)
+	if err != nil {
+		level.Error(log).Log("msg", "FetchChunks", "err", err)
+		return nil, err
+	}
+
+	// Filter out chunks based on the empty matchers in the query.
+	filteredChunks := filterChunksByMatchers(allChunks, allMatchers)
+	return filteredChunks, nil
+}
+
+func (c *seriesStore) GetChunkRefs(ctx context.Context, from, through model.Time, allMatchers ...*labels.Matcher) ([][]Chunk, []*Fetcher, error) {
```

Comment:
Why `[][]Chunk` not `[]Chunk`?  And why `[]*Fetcher` not `*Fetcher`?

Reply:
This is the same interface for `compositeStore`, and there the chunks and the fetchers used to load the chunks are different, ex: schemas that store in Bigtable vs GCS.

So we'll need `[][]Chunk` and `[]*Fetcher` for that and for the same reason we'll need it here too.